### PR TITLE
Allow StorageOrcFileTailSource to read DWRF stripe cache data

### PR DESCRIPTION
**Pull request by sdruzkin**
This change allows StorageOrcFileTailSource to load the DWRF stripe cache data from the source/file into OrcFileTail. This feature is disabled by default. 

To be able to reduce number of IOs while reading the file tail I made the value of the tail read size (the size of the first read operation) configurable.

This is a continuation of https://github.com/prestodb/presto/pull/16102

Test plan:
- added new unit tests for StorageOrcFileTailSource

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:113]: What happens if stripeCacheLength is set to negative or cache mode is set to None.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCacheData.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCacheData.java:28]: The message needs to be "stripeCacheSlice is null".

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:196]: name the dwrfStripeCacheReadSize as dwrfStripeCacheSize.

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:139]: metadataREads from (0, metadataSize), footerReads from (metadataSize, footerSize) and dwrfStripeCacheSlice reads from (0, dwrfStripeCacheSize).

The offset computation looks incorrect.

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:198]: also validate dwrfStripeCacheReadSize >= 0. The reason to check here is we want to only fail when dwrfStripeCacheMode is enabled. so as not to make the current existing code fail.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java:94]: This should be a private method, as it replaces empty with 0 size. If it needs to be a public method, it should return OptionalInt to preserve semantics.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java
Comment (arunthirupathi) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java:103]: should we just call this getTotalSlice.

**Review by arunthirupathi — COMMENTED** [review]
Took a detailed look. Once addressed, this is good to go.

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:113]: Q1: What happens if stripeCacheLength is set to negative?
A1: We would fail down the line, same as if footerSize or metadataSize were set to negative. I can add a bunch of `>0` checks for these 3 fields.

Q2. What happens if cache mode is set to None.
A2: It depends. `None` means that either the cache was not used or that there is a new unsupported cache mode. If the cache was not used and the rest of fields are present we would have an empty cache. If an unsupported mode was used we would and the rest of the cache fields have valid values then we would have an unusable cache.
Do you suggest to proceed only if the cache mode is != NONE?

Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:113]: I did not want to change the code. I like the code as it is.  But want to get your opinions  on it.

Let us made all the new checks only when stripeCacheMode is enabled. If Stripe Cache is corrupted, previously Presto ignored and did not fail. We want it to fail in the new path, but when disabled it should still pass.

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:198]: I think it would make it too strict. Technically `dwrfStripeCacheSize = 0` is a valid scenario even if the `mode is != NONE`- in this case we won't use the cache and fall back to the reading indexes and footers from the stripes.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCacheData.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/DwrfStripeCacheData.java:28]: fixed

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:196]: done

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:139]: Great catch! Fixed + added an assert in the unit test.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java:94]: Made it private. The original purpose was to use it in the `com.facebook.presto.raptor.storage.StorageModule#createOrcFileTailSource`, but I introduced getTotalSlicesSize instead.

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java:103]: I'n not following. Did you meant getTotalSize?

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:198]: Added an independent check to verify `dwrfStripeCacheReadSize >= 0` regardless of metadataSize.

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java:113]: Updated the PR:
- added checks for sizes to not be negative
- avoid loading the stripe cache if mode is NONE

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java
Comment (arunthirupathi) [reply presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcFileTail.java:103]: Sorry, my bad, I mean getTotalSize.

**Review by arunthirupathi — APPROVED** [review]
Looks good, left one minor comment.

**Comment by sdruzkin** [comment]
Renamed `OrcFileTail.getTotalSlicesSize` to `getTotalSize`

File: presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java
Comment (sdruzkin) [reply presto-orc/src/main/java/com/facebook/presto/orc/cache/StorageOrcFileTailSource.java]: The stripe cache size check will be triggered only if this feature is enabled, so we should be good here.

Adding checks for the metadataSize and footerSize just makes these checks more verbose. If I understand it correctly right now these checks are implicitly executed when we attempt to create a slice.

**Comment by sdruzkin** [comment]
@bhhari Can you please route this PR to an appropriate Presto committer?

**Review by bhhari — APPROVED** [review]

**Comment by bhhari** [comment]
@sdruzkin The PR comment is not formatted correct. It should be `=== RELEASE NOTES ===` please check

**Comment by sdruzkin** [comment]
@bhhari Fixed the PR comment. It came from the PR template with two `==`.

**Comment by bhhari** [comment]
> @bhhari Fixed the PR comment. It came from the PR template with two `==`.

I m sorry, I think == is correct https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines. For this PR since these are not major changes, lets just use NO RELEASE NOTES

**Comment by rschlussel** [comment]
@sdruzkin can you explain more about the motivation for "Make ORC read tail size configurable".  I'm not sure it makes sense to include this commit until it's used somewhere.

**Comment by sdruzkin** [comment]
> @sdruzkin can you explain more about the motivation for adding "Make ORC read tail size configurable" as part of this pr. I'm not sure it makes sense to include this commit until it's used somewhere.

The purpose of this series of changes is to reduce IO spent to read a DWRF file. This particular change is used to read the entire tail consisting of stripe cache + footer + post script in one IO operation. Right now the StorageOrcFileTailSource reads last 16KB and this is perfectly fine for most of the files that don't have a lot of stripes and also don't have stripe cache, in this case entire tail would be read using a single IO operation. 

However, when the stripe cache comes into play we would need to increase the default tail read size from 16KB to something around 1MB. 

Here is what we have for some of the existing DWRF files:
Stripe Cache Size:
P50 ~ 25KB
P90 ~ 1.5MB 
P99 ~ 2MB
Max ~ 8.4MB

Footer Size:
P50 ~ 2KB
P90 ~ 21KB
P99 ~ 103KB
P99.99 > 6 MB 

As you can see when stripe cache is enabled with the default 16KB we wouldn't be able to load file tails in one IO even for half of the files. Once support for the stripe cache is fully implemented (I expect 2-3 more PRs) we would immediately enable it in Jolly with a higher value.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java
Comment (rschlussel) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java:43]: static import MEGABYTE

File: presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java
Comment (rschlussel) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java:75]: static import NONE

File: presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java
Comment (rschlussel) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java:104]: static import

File: presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java
Comment (rschlussel) [inline presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java:160]: static import

**Review by rschlussel — COMMENTED** [review]
Just minor style comments - generally we use static imports in Presto unless it would be hard to understand without the fully qualified name.

Otherwise looks good.

File: presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java
Comment (sdruzkin) [reply presto-orc/src/test/java/com/facebook/presto/orc/TestStorageOrcFileTailSource.java]: Done. Replaced all constants and enums in TestStorageOrcFileTailSource with static imports.

**Comment by sdruzkin** [comment]
@rschlussel Do I need to squash commits?

**Review by rschlussel — APPROVED** [review]
Looks good, thanks!  I'll merge once the tests pass

**Comment by sdruzkin** [comment]
All tests are green.

**Pull request merged**
